src/main/java/com/google/googleidentity/oauth2/token/InMemoryOAuth2TokenService.java

```diff
@@ -0,0 +1,451 @@
+                OAuth2AccessToken.newBuilder()
+                        .setAccessToken(info.getTokenString())
+                        .setClientId(request.getRequestAuth().getClientId())
+                        .setUsername(request.getRequestAuth().getUsername())
+                        .setIsScoped(request.getRequestBody().getIsScoped())
+                        .addAllScopes(request.getRequestBody().getScopesList())
+                        .setExpiredTime((new Date().getTime())/1000l + tokenValidTime);
+
+        if (request.getRequestBody().getRefreshable()) {
+            String refreshTokenValue = UUID.randomUUID().toString();
+            UserClientTokenInfo info1 =
+                    new UserClientTokenInfo(username, clientID, refreshTokenValue);
+
+            OAuth2RefreshToken refreshToken =
+                    OAuth2RefreshToken.newBuilder()
+                            .setRefreshToken(info1.getTokenString())
+                            .setClientId(request.getRequestAuth().getClientId())
+                            .setUsername(request.getRequestAuth().getUsername())
+                            .setIsScoped(request.getRequestBody().getIsScoped())
+                            .addAllScopes(request.getRequestBody().getScopesList())
+                            .build();
+
+            userClientRefreshTokenMap.get(username).put(clientID, refreshToken);
+
+            builder.setRefreshToken(refreshToken.getRefreshToken());
+        }
+
+        OAuth2AccessToken token = builder.build();
+
+        accessTokenMap.put(accessTokenValue, token);
+        return token;
+    }
+
+    public Optional<OAuth2AccessToken> refreshToken(String refreshToken) {
+        Optional<OAuth2RefreshToken> token = readRefreshToken(refreshToken);
+
+        if (!token.isPresent()) {
+            return Optional.empty();
+        }
+
+        String username = token.get().getUsername();
+        String clientID = token.get().getClientId();
+
+        HashMap<String, OAuth2AccessToken>  accessTokenMap=
+                    userClientAccessTokenMap.get(username).get(clientID);
+
+        String accessTokenValue = UUID.randomUUID().toString();
+
+        while (accessTokenMap.containsKey(accessTokenValue)) {
+            accessTokenValue = UUID.randomUUID().toString();
+        }
+
+        UserClientTokenInfo info =
+                new UserClientTokenInfo(username, clientID, accessTokenValue);
+
+        OAuth2AccessToken newToken =
+                OAuth2AccessToken.newBuilder()
+                        .setAccessToken(info.getTokenString())
+                        .setClientId(clientID)
+                        .setUsername(username)
+                        .setIsScoped(token.get().getIsScoped())
+                        .addAllScopes(token.get().getScopesList())
+                        .setExpiredTime((new Date().getTime())/1000l + tokenValidTime)
+                        .setRefreshToken(refreshToken)
+                        .build();
```

Comment:
This looks similar to generateToken. Can the common logic to generate a valid access token be broken out and shared between these two methods?